### ISSUE TITLE
Occasionally getting Timeout expired on second BulkInsert operation

### ISSUE THREAD

**Issue by ddavidtran** [borisdj/EFCore.BulkExtensions#190]
Hi,

Had previously an issue which was recently closed: [#176](https://github.com/borisdj/EFCore.BulkExtensions/issues/176)

I'm creating a new issue as I have not solved this yet/put it behind me since if all insert operation works then I'm basically content, which is about 50% of the time..

I'm using this extension to populate a database with nested entity relationships. 

My models:

__Student__:
```cs
public class Student
{
 public int StudentId {get;set;}
 public string StudentName {get;set;}
 public List<Course> Course{ get; } = new List<Course>();
}
```

__Course__:
```cs
public class Course
{
 public int CourseId {get;set;}
 public string CourseName {get;set;}

 public int StudentId {get;set;}
 public Student Student {get;set;}

 public List<Grade> Grades{ get; } = new List<Grade>();
}
```

__Grade__:
```cs
public class Grade
{
 public int GradeId {get;set;}
 public string GradeValue{get;set;}

 public int CourseId {get;set;}
 public Course Course {get;set;}
}
```

Student has a _one-to-many_ relationship with Course.

Course has a _one-to-many_ relationship with Grade.

The problem I get is that I get _**occasionally**_ timeout expired error for one specific BulkInsert operation. See code below:

```cs
int i = 0;
students.ForEach(s => s.StudentId = ++i);
using (var transaction = context.Database.BeginTransaction())
{
 await context.BulkInsertAsync(students, new BulkConfig { SetOutputIdentity = true, PreserveInsertOrder = true});

 int j = 0; 
 foreach(var student in students)
 {
   foreach (var course in student.Courses)
   { 
     course.CourseId = ++j; 
     course.StudentId= student.StudentId;
   }
 }

var courses = students.SelectMany(s => s.Course).ToList();
await context.BulkInsertAsync(courses, new BulkConfig{ SetOutputIdentity = true, PreserveInsertOrder=true }); //Occasionally timeouts, ONLY for this insertion. All insertion operations are successful IF it passes this line,

 foreach(var course in courses)
 {
   foreach (var grade in course.Grades)
   { 
     grade.CourseId= course.CourseId;
   }
 }
var grades= courses.SelectMany(c => c.Grades).ToList();

await context.BulkInsertAsync(grades); 
transaction.Commit()
}
```

The problem line is `await context.BulkInsertAsync(courses, new BulkConfig{ SetOutputIdentity = true, PreserveInsertOrder=true });`
This is very weird. There are no problems with other insertion operation, for instance BulkInsert operation for students works **everytime**, and if the insertion for courses are successful, then the BulkInsert for grades works **everytime** as well. The only problem is when I insert courses. I can't seem to pinpoint the exact cause of it. It happens very randomly, sometimes three times in a row, sometimes after four successful insertions. 

**tl:dr** I'm getting timeout expired error occasionally for a specific BulkInsertion, all other bulkinsert works. Can't seem to pinpoint the exact cause of this.

Thanks,

**Comment by borisdj** [comment]
What I can notice is that since you are using _async_ methods Transaction should also be async, so:
`using (var transaction = await context.Database.BeginTransactionAsync())`

**Comment by ddavidtran** [comment]
> What I can notice is that since you are using _async_ methods Transaction should also be async, so:
> `using (var transaction = await context.Database.BeginTransactionAsync())`

Before I had transaction and all insertions synchronous. I recently switched to async and I still get occassionally timeouts on that specific insert..

**Comment by borisdj** [comment]
I am not able to reproduce this.
Anyway you could trt change BulkCopyTimeout to higher value or no limit:

> BulkCopyTimeout when not set has SqlBulkCopy default which is 30 seconds and if set to 0 it indicates no limit.

**Comment by provalinf** [comment]
I am also facing this problem during the second import only on my mysql azure server. Locally no problem. It's quite frustrating, I needed it for a deployment, I will try to find an alternative :/
```
await using (var transaction = await _dbContext.Database.BeginTransactionAsync())
{
	await _dbContext.BulkInsertAsync(resultImport.Item1, new BulkConfig { BatchSize = 1000, BulkCopyTimeout = 0 });
	await _dbContext.BulkInsertAsync(resultImport.Item2, new BulkConfig { BatchSize = 20000, BulkCopyTimeout = 0});
	await transaction.CommitAsync();
}
```
